From 6e4d2556b6388b17fe52a30df3547493c9ba5500 Mon Sep 17 00:00:00 2001
From: Jason Rogena <j.rogena@cgiar.org>
Date: Fri, 13 Jun 2014 12:46:55 +0300
Subject: [PATCH 196/213] Version 1.0.1: Refresing farmers, farmer filtering

Farmer data is now refreshed after an edit in EditFarmerActivity

Farmer list can now be filtered based on the extension personnel
tied to that farmer in FarmerSelectioActivity
---
 .../org/cgiar/ilri/mistro/farmer/EditFarmer.java   |   2 +-
 .../cgiar/ilri/mistro/farmer/FarmerSelection.java  | 212 ++++++++++++++++++---
 .../org/cgiar/ilri/mistro/farmer/MainMenu.java     |   7 +-
 .../cgiar/ilri/mistro/farmer/carrier/Farmer.java   |   5 +-
 .../main/res/layout/activity_farmer_selection.xml  |  20 +-
 MistroFarmer/src/main/res/values/strings_en.xml    |   5 +
 6 files changed, 217 insertions(+), 34 deletions(-)

diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/EditFarmer.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/EditFarmer.java
index 83f4256..d38673c 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/EditFarmer.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/EditFarmer.java
@@ -548,7 +548,7 @@ public class EditFarmer extends SherlockActivity implements MistroActivity, View
                 Toast.makeText(EditFarmer.this, Locale.getStringInLocale("farmer_profile_updated", EditFarmer.this), Toast.LENGTH_LONG).show();
 
                 Intent intent = new Intent(EditFarmer.this, FarmerSelection.class);
-                intent.putExtra(FarmerSelection.KEY_ADMIN_DATA, adminData);
+                //intent.putExtra(FarmerSelection.KEY_ADMIN_DATA, adminData);
                 startActivity(intent);
             }
         }
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerSelection.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerSelection.java
index 1584ac6..5ff0fbc 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerSelection.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerSelection.java
@@ -1,48 +1,63 @@
 package org.cgiar.ilri.mistro.farmer;
 
+import android.app.ProgressDialog;
+import android.content.Context;
 import android.content.Intent;
+import android.os.AsyncTask;
 import android.os.Bundle;
+import android.telephony.TelephonyManager;
 import android.util.Log;
 import android.view.View;
+import android.widget.AdapterView;
 import android.widget.ArrayAdapter;
 import android.widget.Button;
 import android.widget.Spinner;
 import android.widget.TextView;
+import android.widget.Toast;
 
 import com.actionbarsherlock.app.SherlockActivity;
 import com.actionbarsherlock.view.Menu;
 import com.actionbarsherlock.view.MenuInflater;
 import com.actionbarsherlock.view.MenuItem;
 
+import org.cgiar.ilri.mistro.farmer.backend.DataHandler;
 import org.cgiar.ilri.mistro.farmer.backend.Locale;
 import org.cgiar.ilri.mistro.farmer.carrier.Farmer;
 import org.json.JSONArray;
+import org.json.JSONException;
 import org.json.JSONObject;
 
 import java.util.ArrayList;
 import java.util.List;
 
 
-public class FarmerSelection extends SherlockActivity implements MistroActivity, View.OnClickListener {
+public class FarmerSelection extends SherlockActivity implements MistroActivity, View.OnClickListener, AdapterView.OnItemSelectedListener {
 
     private static final String TAG = "FarmerSelection";
     public static final String KEY_ADMIN_DATA= "adminData";
 
     private Menu menu;
 
+    private TextView filterFarmersTV;
+    private Spinner filterFarmersS;
     private TextView selectFarmerTV;
     private Spinner selectFarmerS;
     private Button selectB;
     private Button backB;
 
-    private List<Farmer> farmers;
+    private List<Farmer> allFarmers;
+    private List<Farmer> filteredFarmers;
     private JSONObject adminData;
+    private List<String> filters;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_farmer_selection);
 
+        filterFarmersTV = (TextView)findViewById(R.id.filter_farmers_tv);
+        filterFarmersS = (Spinner)findViewById(R.id.filter_farmers_s);
+        filterFarmersS.setOnItemSelectedListener(this);
         selectFarmerTV = (TextView)findViewById(R.id.select_farmer_tv);
         selectFarmerS = (Spinner)findViewById(R.id.select_farmer_s);
 
@@ -61,33 +76,73 @@ public class FarmerSelection extends SherlockActivity implements MistroActivity,
         Bundle bundle=this.getIntent().getExtras();
         if(bundle != null){
             String adminJSONString = bundle.getString(KEY_ADMIN_DATA);
-            try{
-                adminData = new JSONObject(adminJSONString);
-                JSONArray farmerData = adminData.getJSONArray("farmers");
-
-                farmers = new ArrayList<Farmer>(farmerData.length());
-                for(int i = 0; i < farmerData.length(); i++){
-                    String ePersonnel = "";
-                    if(!farmerData.getJSONObject(i).getString("extension_personnel_id").equals("NULL")){
-                        ePersonnel = adminData.getString("name");
-                    }
-                    Farmer currFarmer = new Farmer(farmerData.getJSONObject(i), ePersonnel);
-                    farmers.add(currFarmer);
-                }
+            loadAdminData(adminJSONString);
+        }
+        else{
+            TelephonyManager telephonyManager=(TelephonyManager)this.getSystemService(Context.TELEPHONY_SERVICE);
+            if(telephonyManager != null){
+                GetFarmerDataThread getFarmerDataThread = new GetFarmerDataThread();
+                getFarmerDataThread.execute(telephonyManager.getSimSerialNumber());
+            }
+            else{
+                Toast.makeText(this,Locale.getStringInLocale("no_sim_card",this),Toast.LENGTH_LONG).show();
+            }
+        }
+    }
 
-                List<String> farmerNames = new ArrayList<String>(farmers.size());
-                for(int i = 0; i < farmers.size(); i++){
-                    Farmer currFarmer = farmers.get(i);
-                    farmerNames.add(currFarmer.getFullName());
-                }
+    private void loadAdminData(String adminJSONString){
+        try {
+            adminData = new JSONObject(adminJSONString);
+            JSONArray farmerData = adminData.getJSONArray("farmers");
 
-                ArrayAdapter<String> farmerArrayAdapter = new ArrayAdapter<String>(this,android.R.layout.simple_spinner_item, farmerNames);
-                farmerArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
-                selectFarmerS.setAdapter(farmerArrayAdapter);
+            filters = new ArrayList<String>();
+            filters.add(Locale.getStringInLocale("all_farmers", this));
+            filters.add(Locale.getStringInLocale("farmers_no_epersonnel", this));
+            if(adminData.getInt("is_super") == 1){//admin is super
+                JSONArray allEPersonnel = adminData.getJSONArray("extension_personnel");
+                for(int i = 0; i < allEPersonnel.length(); i++){
+                    JSONObject currEPersonnel = allEPersonnel.getJSONObject(i);
+                    filters.add(Locale.getStringInLocale("farmers_tied_to", this) + " " + currEPersonnel.getString("name"));
+                }
             }
-            catch (Exception e){
+            else{
+                filters.add(Locale.getStringInLocale("farmers_tied_to", this) +" "+ adminData.getString("name"));
+            }
+            ArrayAdapter<String> filterArrayAdapter = new ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, filters);
+            filterArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+            filterFarmersS.setAdapter(filterArrayAdapter);
 
+            allFarmers = new ArrayList<Farmer>(farmerData.length());
+            filteredFarmers = new ArrayList<Farmer>(farmerData.length());
+            for (int i = 0; i < farmerData.length(); i++) {
+                String ePersonnel = "";
+                if (!farmerData.getJSONObject(i).getString("extension_personnel_id").equals("NULL")) {
+                    ePersonnel = adminData.getString("name");
+                }
+                Farmer currFarmer = new Farmer(farmerData.getJSONObject(i));
+                allFarmers.add(currFarmer);
+                filteredFarmers.add(currFarmer);
             }
+
+            setFilteredFarmerList(filteredFarmers);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private void setFilteredFarmerList(List<Farmer> filteredFarmers){
+        List<String> farmerNames = new ArrayList<String>(filteredFarmers.size());
+        for (int i = 0; i < filteredFarmers.size(); i++) {
+            Farmer currFarmer = filteredFarmers.get(i);
+            farmerNames.add(currFarmer.getFullName());
+        }
+
+        ArrayAdapter<String> farmerArrayAdapter = new ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, farmerNames);
+        farmerArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+        selectFarmerS.setAdapter(farmerArrayAdapter);
+
+        if(this.filteredFarmers != filteredFarmers) {
+            this.filteredFarmers = filteredFarmers;
         }
     }
 
@@ -117,6 +172,7 @@ public class FarmerSelection extends SherlockActivity implements MistroActivity,
     public void initTextInViews() {
         setTitle(Locale.getStringInLocale("select_farmer", this));
 
+        filterFarmersTV.setText(Locale.getStringInLocale("filter_farmers", this));
         selectFarmerTV.setText(Locale.getStringInLocale("select_farmer", this));
         selectB.setText(Locale.getStringInLocale("select", this));
         backB.setText(Locale.getStringInLocale("back", this));
@@ -132,9 +188,9 @@ public class FarmerSelection extends SherlockActivity implements MistroActivity,
     @Override
     public void onClick(View v) {
         if(v.equals(selectB)){
-            if(selectFarmerS.getSelectedItemPosition() != -1 && farmers.size() > selectFarmerS.getSelectedItemPosition()){
+            if(selectFarmerS.getSelectedItemPosition() != -1 && filteredFarmers.size() > selectFarmerS.getSelectedItemPosition()){
                 Log.d(TAG, "Selected farmer index = "+String.valueOf(selectFarmerS.getSelectedItemPosition()));
-                Farmer selectedFarmer = farmers.get(selectFarmerS.getSelectedItemPosition());
+                Farmer selectedFarmer = filteredFarmers.get(selectFarmerS.getSelectedItemPosition());
                 Intent intent = new Intent(this, EditFarmer.class);
                 Bundle bundle = new Bundle();
                 bundle.putParcelable(Farmer.PARCELABLE_KEY, selectedFarmer);
@@ -150,4 +206,108 @@ public class FarmerSelection extends SherlockActivity implements MistroActivity,
             startActivity(intent);
         }
     }
+
+    @Override
+    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
+        if(parent == filterFarmersS){
+            if(filterFarmersS.getSelectedItemPosition() == 0){//all farmers
+                List<Farmer> newlyFilteredFarmers = this.allFarmers;
+                setFilteredFarmerList(newlyFilteredFarmers);
+            }
+            else if(filterFarmersS.getSelectedItemPosition() == 1){//farmers without extension personnel
+                List<Farmer> newlyFilteredFarmers = new ArrayList<Farmer>();
+                for(int i =0; i<allFarmers.size(); i++){
+                    Farmer currFarmer = allFarmers.get(i);
+                    if(DataHandler.isNull(currFarmer.getExtensionPersonnel()) || currFarmer.getExtensionPersonnel().length() == 0){
+                        newlyFilteredFarmers.add(currFarmer);
+                    }
+                    else{
+                        Log.d(TAG, currFarmer.getExtensionPersonnel()+" does not qualify to be null");
+                    }
+                }
+
+                setFilteredFarmerList(newlyFilteredFarmers);
+            }
+            else if(filterFarmersS.getSelectedItemPosition() != -1){
+                String selection = filters.get(filterFarmersS.getSelectedItemPosition());
+
+                List<Farmer> newlyFilteredFarmers = new ArrayList<Farmer>();
+                for(int i = 0; i < allFarmers.size(); i++){
+                    Farmer currFarmer = allFarmers.get(i);
+                    if(currFarmer.getExtensionPersonnel() != null && currFarmer.getExtensionPersonnel().length() > 0 &&
+                            selection.contains(currFarmer.getExtensionPersonnel())){
+                        newlyFilteredFarmers.add(currFarmer);
+                    }
+                    else{
+                        Log.d(TAG, currFarmer.getExtensionPersonnel()+" does not match "+selection);
+                    }
+                }
+
+                setFilteredFarmerList(newlyFilteredFarmers);
+            }
+        }
+    }
+
+    @Override
+    public void onNothingSelected(AdapterView<?> parent) {
+
+    }
+
+    private class GetFarmerDataThread extends AsyncTask<String, Integer, String>{
+
+        private ProgressDialog progressDialog;
+
+        @Override
+        protected void onPreExecute() {
+            super.onPreExecute();
+            progressDialog = ProgressDialog.show(FarmerSelection.this, "", Locale.getStringInLocale("loading_please_wait", FarmerSelection.this), true);
+        }
+
+        @Override
+        protected String doInBackground(String... params) {
+            JSONObject jsonObject=new JSONObject();
+            try
+            {
+                jsonObject.put("simCardSN",params[0]);
+                jsonObject.put("deviceType", "Android");
+                //jsonObject.put("mobileNumber",params[1]);
+                String result = DataHandler.sendDataToServer(FarmerSelection.this, jsonObject.toString(), DataHandler.ADMIN_AUTHENTICATION_URL, true);
+                return result;
+            }
+            catch (JSONException e)
+            {
+                e.printStackTrace();
+            }
+            return null;
+        }
+
+        @Override
+        protected void onPostExecute(String result) {
+            super.onPostExecute(result);
+            progressDialog.dismiss();
+            if(result==null){
+                String httpError = DataHandler.getSharedPreference(FarmerSelection.this, "http_error", "No Error thrown to application. Something must be really wrong");
+                Toast.makeText(FarmerSelection.this, httpError, Toast.LENGTH_LONG).show();
+            }
+            else if(result.equals(DataHandler.SMS_ERROR_GENERIC_FAILURE)){
+                Toast.makeText(FarmerSelection.this, Locale.getStringInLocale("generic_sms_error", FarmerSelection.this), Toast.LENGTH_LONG).show();
+            }
+            else if(result.equals(DataHandler.SMS_ERROR_NO_SERVICE)){
+                Toast.makeText(FarmerSelection.this, Locale.getStringInLocale("no_service", FarmerSelection.this), Toast.LENGTH_LONG).show();
+            }
+            else if(result.equals(DataHandler.SMS_ERROR_RADIO_OFF)){
+                Toast.makeText(FarmerSelection.this, Locale.getStringInLocale("radio_off", FarmerSelection.this), Toast.LENGTH_LONG).show();
+            }
+            else if(result.equals(DataHandler.SMS_ERROR_RESULT_CANCELLED)){
+                Toast.makeText(FarmerSelection.this, Locale.getStringInLocale("server_not_receive_sms", FarmerSelection.this), Toast.LENGTH_LONG).show();
+            }
+            else if(result.equals(DataHandler.CODE_USER_NOT_AUTHENTICATED)){
+                Log.w(TAG, "Admin not authenticated. May mean he/she changed sim cards after logging in");
+                Toast.makeText(FarmerSelection.this, Locale.getStringInLocale("sim_card_not_admin", FarmerSelection.this), Toast.LENGTH_LONG).show();
+            }
+            else{
+                loadAdminData(result);
+            }
+        }
+    }
 }
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MainMenu.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MainMenu.java
index 9435d2f..3788215 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MainMenu.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MainMenu.java
@@ -102,8 +102,11 @@ public class MainMenu extends SherlockActivity implements MistroActivity, View.O
                     Log.d(TAG, " *** Admin data from login screen = "+adminJSONString);
                     try{
                         adminData = new JSONObject(adminJSONString);
-
-                        Toast.makeText(this, Locale.getStringInLocale("welcome", this) + " " + adminData.getString("name") + " (" + Locale.getStringInLocale("admin", this) + ")", Toast.LENGTH_LONG).show();
+                        String type = Locale.getStringInLocale("admin", this);
+                        if(adminData.getInt("is_super") == 1){
+                            type = Locale.getStringInLocale("super_admin", this);
+                        }
+                        Toast.makeText(this, Locale.getStringInLocale("welcome", this) + " " + adminData.getString("name") + " (" + type + ")", Toast.LENGTH_LONG).show();
                     }
                     catch (Exception e){
                         e.printStackTrace();
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java
index 236459c..ca67df6 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java
@@ -4,6 +4,7 @@ import android.os.Parcel;
 import android.os.Parcelable;
 import android.util.Log;
 
+import org.cgiar.ilri.mistro.farmer.backend.DataHandler;
 import org.json.JSONArray;
 import org.json.JSONException;
 import org.json.JSONObject;
@@ -56,12 +57,12 @@ public class Farmer implements Parcelable, Serializable
         readFromParcel(source);
     }
 
-    public Farmer(JSONObject farmerJsonObject, String extensionPersonnel){
+    public Farmer(JSONObject farmerJsonObject){
         try{
             Log.d(TAG, "Json for this farmer = "+farmerJsonObject.toString());
             id = farmerJsonObject.getInt("id");
             fullName = farmerJsonObject.getString("name");
-            this.extensionPersonnel = extensionPersonnel;
+            this.extensionPersonnel = (DataHandler.isNull(farmerJsonObject.getString("extension_personnel"))) ? "" : farmerJsonObject.getString("extension_personnel");
             mobileNumber = farmerJsonObject.getString("mobile_no");
             longitude = farmerJsonObject.getString("gps_longitude");
             latitude = farmerJsonObject.getString("gps_latitude");
diff --git a/MistroFarmer/src/main/res/layout/activity_farmer_selection.xml b/MistroFarmer/src/main/res/layout/activity_farmer_selection.xml
index 4604465..393de3a 100644
--- a/MistroFarmer/src/main/res/layout/activity_farmer_selection.xml
+++ b/MistroFarmer/src/main/res/layout/activity_farmer_selection.xml
@@ -10,14 +10,28 @@
         android:paddingRight="@dimen/activity_horizontal_margin"
         android:paddingTop="@dimen/activity_vertical_margin"
         android:paddingBottom="@dimen/activity_vertical_margin"
-        tools:context=".MainMenu">
+        tools:context=".FarmerSelection">
         <TextView
-            android:id="@+id/select_farmer_tv"
+            android:id="@+id/filter_farmers_tv"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/activity_f_r_text_view_margin_top"
             android:textSize="@dimen/normal_text_size"/>
+        <Spinner
+            android:id="@+id/filter_farmers_s"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/activity_f_r_edit_text_margin_top"
+            android:layout_marginLeft="@dimen/activity_f_r_edit_text_margin_left"
+            android:textSize="@dimen/normal_text_size"
+            android:textColor="@color/text_input_color" />
 
-
+        <TextView
+            android:id="@+id/select_farmer_tv"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/activity_f_r_text_view_margin_top"
+            android:textSize="@dimen/normal_text_size"/>
         <Spinner
             android:id="@+id/select_farmer_s"
             android:layout_width="match_parent"
diff --git a/MistroFarmer/src/main/res/values/strings_en.xml b/MistroFarmer/src/main/res/values/strings_en.xml
index 326be99..ce9148d 100644
--- a/MistroFarmer/src/main/res/values/strings_en.xml
+++ b/MistroFarmer/src/main/res/values/strings_en.xml
@@ -329,4 +329,9 @@
     <string name="are_you_in_farmers_farm_en">Are you currently in the farmer\'s farm?</string>
     <string name="accuracy_en">Accuracy</string>
     <string name="unknown_en">unknown</string>
+    <string name="filter_farmers_en">Filter farmers</string>
+    <string name="all_farmers_en">All farmers</string>
+    <string name="farmers_no_epersonnel_en">Farmers with no extension personnel</string>
+    <string name="farmers_tied_to_en">Farmers tied to</string>
+    <string name="super_admin_en">Super Admin</string>
 </resources>
-- 
2.7.1

